client/jetpack-connect/controller.js

```diff
@@ -177,6 +178,20 @@ export function connect( context, next ) {
 	next();
 }
 
+export function instructions( context, next ) {
+	analytics.pageView.record(
+		'jetpack/connect/instructions',
+		'Jetpack Manual Install Instructions'
+	);
+
+	const url = context.query.url;
+	if ( ! url ) {
```

Comment:
Good question. Redirects from the controller are preferable because they are much simpler. But they can only happen once, so can only be done _if all the necessary data is guaranteed to be available_. 

For example, in the above case, we redirect based on the absence of a query param in the current URL. We know for sure that that piece of data is available when we hit the controller.

In the credentials form, we are [redirecting based on redux state](https://github.com/Automattic/wp-calypso/blob/master/client/jetpack-connect/remote-credentials.js#L63), which generally does not become available until some network request completes, meaning we have to wait for props to update inside the component to be safe.

As it happens, that particular piece of redux state (remote site url) should always be available because it is entered by the user before we arrive, so we could do that redirect in the controller. I would favour adding the remote url as a query on the install route, just as we have done with the instructions, making it clear that it should always be available.

Reply:
Thanks for elaborating 👍